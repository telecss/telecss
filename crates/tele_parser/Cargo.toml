[package]
name = "tele_parser"
version = "0.2.0"
description = "A CSS parser"
edition = "2021"
license = "MIT"
repository = "https://github.com/telecss/telecss"
keywords = ["css", "telecss", "tele_parser", "parser"]

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
tele_tokenizer = { path = "../tele_tokenizer", version = "0.2.0" }
serde = { version = "1.0.133", features = ["derive", "rc"] }

[dev-dependencies]
insta = "1.10.0"