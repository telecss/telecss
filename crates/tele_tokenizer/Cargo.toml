[package]
name = "tele_tokenizer"
version = "0.1.0"
description = "A CSS tokenizer"
edition = "2021"
repository = "https://github.com/telecss/telecss"
license = "MIT"
keywords = ["css", "telecss", "tele_tokenizer", "tokenizer"]

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
tele_utils = { path = "../tele_utils", version = "0.1.0" }
serde = { version = "1.0.133", features = ["derive"] }

[dev-dependencies]
insta = "1.10.0"

[features]
