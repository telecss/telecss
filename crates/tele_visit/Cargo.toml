[package]
name = "tele_visit"
version = "0.1.0"
description = "A CSS AST visitor for visit the AST produced by tele_parser"
edition = "2021"
repository = "https://github.com/telecss/telecss"
license = "MIT"
keywords = ["css", "telecss", "tele_visit", "visit"]

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
tele_parser = { path = "../tele_parser", version = "0.1.0" }

[dev-dependencies]
insta = "1.10.0"
tele_tokenizer = { path = "../tele_tokenizer", version = "0.1.0" }