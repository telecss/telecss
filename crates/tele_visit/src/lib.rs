#![warn(missing_docs)]
#![deny(unused_imports)]
#![deny(unused_variables)]

//! Visitor/Folder pattern for transforming the AST

use std::{cell::RefCell, rc::Rc};
use tele_parser::{
  AtRuleNode, DeclarationNode, DimensionNode, FunctionNode, IdentNode, NumberNode, OperatorNode,
  PercentageNode, RawNode, RuleSetNode, StatementNode, StringNode, StyleSheetNode, URLNode, Value,
};

pub trait MutVisitor<'s> {
  fn visit_ss_node(&self, _ss_node: Rc<RefCell<StyleSheetNode<'s>>>) {}
  fn visit_rule_set_node(&self, _rule_set_node: Rc<RefCell<RuleSetNode<'s>>>) {}
  fn visit_at_rule_node(&self, _at_rule_node: Rc<RefCell<AtRuleNode<'s>>>) {}
  fn visit_decl_node(&self, _decl_node: Rc<RefCell<DeclarationNode<'s>>>) {}
  fn visit_ident_node(&self, _decl_node: Rc<RefCell<IdentNode>>) {}
  fn visit_url_node(&self, _decl_node: Rc<RefCell<URLNode>>) {}
  fn visit_dimension_node(&self, _decl_node: Rc<RefCell<DimensionNode>>) {}
  fn visit_number_node(&self, _decl_node: Rc<RefCell<NumberNode>>) {}
  fn visit_operator_node(&self, _decl_node: Rc<RefCell<OperatorNode>>) {}
  fn visit_percentage_node(&self, _decl_node: Rc<RefCell<PercentageNode>>) {}
  fn visit_string_node(&self, _decl_node: Rc<RefCell<StringNode>>) {}
  fn visit_fn_node(&self, _decl_node: Rc<RefCell<FunctionNode>>) {}
  fn visit_raw_node(&self, _decl_node: Rc<RefCell<RawNode>>) {}
}

pub struct VisitMut<'s> {
  ast: Rc<RefCell<StyleSheetNode<'s>>>,
  visitors: Vec<Box<dyn MutVisitor<'s>>>,
}

impl<'s> VisitMut<'s> {
  pub fn new(ast: Rc<RefCell<StyleSheetNode<'s>>>, visitors: Vec<Box<dyn MutVisitor<'s>>>) -> Self {
    Self { ast, visitors }
  }

  pub fn transform(&self) {
    self.visit_ss_node(Rc::clone(&self.ast));
  }

  /// Visit [StyleSheetNode] node
  fn visit_ss_node(&self, ss_node: Rc<RefCell<StyleSheetNode<'s>>>) {
    for visitor in self.visitors.iter() {
      visitor.visit_ss_node(Rc::clone(&ss_node));
    }

    for stat in ss_node.borrow().statements.iter() {
      match stat {
        StatementNode::AtRule(node) => self.visit_at_rule_node(Rc::clone(node)),
        StatementNode::RuleSet(node) => self.visit_rule_set_node(Rc::clone(node)),
      }
    }
  }

  /// Visit [RuleSetNode] node
  fn visit_rule_set_node(&self, rule_set_node: Rc<RefCell<RuleSetNode<'s>>>) {
    for visitor in self.visitors.iter() {
      visitor.visit_rule_set_node(Rc::clone(&rule_set_node));
    }

    for decl in rule_set_node.borrow().declarations.iter() {
      self.visit_decl_node(Rc::clone(decl))
    }
  }

  /// Visit [AtRuleNode] node
  fn visit_at_rule_node(&self, at_rule_node: Rc<RefCell<AtRuleNode<'s>>>) {
    for visitor in self.visitors.iter() {
      visitor.visit_at_rule_node(Rc::clone(&at_rule_node));
    }

    for stat in at_rule_node.borrow().block.iter() {
      match stat {
        StatementNode::AtRule(node) => self.visit_at_rule_node(Rc::clone(node)),
        StatementNode::RuleSet(node) => self.visit_rule_set_node(Rc::clone(node)),
      }
    }
  }

  /// Visit [DeclarationNode] node
  fn visit_decl_node(&self, decl_node: Rc<RefCell<DeclarationNode<'s>>>) {
    for visitor in self.visitors.iter() {
      visitor.visit_decl_node(Rc::clone(&decl_node));
    }

    for value in decl_node.borrow().value.iter() {
      match value {
        Value::Ident(node) => {
          for visitor in self.visitors.iter() {
            visitor.visit_ident_node(Rc::clone(node));
          }
        }
        Value::URL(node) => {
          for visitor in self.visitors.iter() {
            visitor.visit_url_node(Rc::clone(node));
          }
        }
        Value::Dimension(node) => {
          for visitor in self.visitors.iter() {
            visitor.visit_dimension_node(Rc::clone(node));
          }
        }
        Value::Number(node) => {
          for visitor in self.visitors.iter() {
            visitor.visit_number_node(Rc::clone(node));
          }
        }
        Value::Operator(node) => {
          for visitor in self.visitors.iter() {
            visitor.visit_operator_node(Rc::clone(node));
          }
        }
        Value::Percentage(node) => {
          for visitor in self.visitors.iter() {
            visitor.visit_percentage_node(Rc::clone(node));
          }
        }
        Value::String(node) => {
          for visitor in self.visitors.iter() {
            visitor.visit_string_node(Rc::clone(node));
          }
        }
        Value::Function(node) => {
          for visitor in self.visitors.iter() {
            visitor.visit_fn_node(Rc::clone(node));
          }
        }
        Value::Raw(node) => {
          for visitor in self.visitors.iter() {
            visitor.visit_raw_node(Rc::clone(node));
          }
        }
      };
    }
  }
}
