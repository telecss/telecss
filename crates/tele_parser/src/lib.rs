#![warn(missing_docs)]
#![deny(unused_imports)]
#![deny(unused_variables)]

//! A simpler and faster CSS parser.
//!
//! # Usages
//! ```rust
//! use tele_tokenizer::Tokenizer;
//! use tele_parser::Parser;
//!
//! // Create a tokenizer
//! let mut tokenizer: Tokenizer = r".foo { color: red; }".into();
//! // Tokenize based on the given raw string,
//! let tokens = tokenizer.tokenize().unwrap();
//! // Create a parser from a sequence of tokens
//! let parser = Parser::from(tokens);
//! // parsing it
//! let ast = parser.parse();
//! ```

mod error;
mod node;

pub use error::Result as ParserResult;
use error::{SyntaxError, SyntaxErrorKind};
pub use node::*;

use std::{cell::RefCell, iter::Peekable, rc::Rc, slice::Iter};

use tele_tokenizer::*;

/// Represents the CSS parser.
pub struct Parser<'s> {
  iter: RefCell<Peekable<Iter<'s, Token<'s>>>>,
  value_fncall_stack: RefCell<Vec<u8>>,
}

impl<'s> From<&'s Vec<Token<'s>>> for Parser<'s> {
  fn from(tokens: &'s Vec<Token<'s>>) -> Self {
    Self {
      iter: RefCell::new(tokens.iter().peekable()),
      value_fncall_stack: RefCell::new(vec![]),
    }
  }
}

impl<'s> Parser<'s> {
  /// Perform parsing
  pub fn parse(&self) -> ParserResult<AstType> {
    let mut ss_node = StyleSheetNode::default();

    while let Some(node) = self.parse_statements()? {
      ss_node.statements.push(node)
    }

    self.skip_ws_and_comments();

    let next = self.peek();
    debug_assert!(next.is_eof());

    ss_node.loc.end = next.end_pos;

    Ok(Rc::new(RefCell::new(ss_node)))
  }

  fn peek(&self) -> &Token {
    let next = self.iter.borrow_mut().peek().map(|&token| token);
    // Because the last token will always be EOF2, and we will handle it reasonably.
    // So in general/theoretically, we don't encounter `None`.
    debug_assert!(next.is_some());
    unsafe { next.unwrap_unchecked() }
  }

  fn consume(&self) -> Option<&Token> {
    self.iter.borrow_mut().next()
  }

  fn parse_statements(&self) -> ParserResult<Option<StatementNode>> {
    self.skip_ws_and_comments();

    let next = self.peek();

    if next.is_eof() {
      return Ok(None);
    }

    if next.is_at_keyword() {
      Ok(
        self
          .parse_at_rule()?
          .map(|node| StatementNode::AtRule(Rc::new(RefCell::new(node)))),
      )
    } else {
      Ok(
        self
          .parse_rule_set()?
          .map(|node| StatementNode::RuleSet(Rc::new(RefCell::new(node)))),
      )
    }
  }

  fn parse_rule_set(&self) -> ParserResult<Option<RuleSetNode>> {
    let mut rule_set_node = RuleSetNode::default();

    let mut token = self.peek();
    rule_set_node.loc.start = token.start_pos;

    loop {
      token = self.peek();
      self.asset_eof(token)?;

      if token.is_lcb() {
        self.consume();
        while let Some(node) = self.parse_decl()? {
          rule_set_node.declarations.push(Rc::new(RefCell::new(node)));
        }
      } else if token.is_rcb() {
        self.consume();
        break;
      } else {
        if !token.is_comment() {
          rule_set_node.prelude.push_str(&token.to_string());
          rule_set_node.prelude_tokens.push(token);
        }
        self.consume();
      }
    }

    rule_set_node.prelude = rule_set_node.prelude.trim_end().to_string();
    rule_set_node.loc.end = token.end_pos;

    Ok(Some(rule_set_node))
  }

  fn parse_at_rule(&self) -> ParserResult<Option<AtRuleNode>> {
    let mut at_rule_node = AtRuleNode::default();

    // unsafe: we are sure that the next token is `@`
    let mut token = unsafe { self.consume().unwrap_unchecked() };
    at_rule_node.loc.start = token.start_pos;
    at_rule_node.name = token.to_string();
    at_rule_node.name_tokens.push(token);

    self.skip_ws_and_comments();

    loop {
      token = self.peek();
      self.asset_eof(token)?;

      if token.is_lcb() {
        self.consume();

        loop {
          self.skip_ws_and_comments();
          let next = self.peek();
          if next.is_rcb() {
            // end right curly bracket
            self.consume();
            break;
          }

          let statement = self.parse_statements()?;
          if statement.is_none() {
            break;
          }

          // unsafe: checked above
          let statement = unsafe { statement.unwrap_unchecked() };

          at_rule_node.block.push(statement);
        }
        break;
      } else if token.is_semi_colon() {
        // no-block
        self.consume();
        break;
      } else {
        // prelude
        self.consume();
        at_rule_node.prelude.push_str(&token.to_string());
        at_rule_node.prelude_tokens.push(token);
      }
    }

    at_rule_node.loc.end = token.end_pos;
    at_rule_node.prelude = at_rule_node.prelude.trim_end().to_string();

    Ok(Some(at_rule_node))
  }

  fn parse_decl(&self) -> ParserResult<Option<DeclarationNode>> {
    self.skip_ws_and_comments();

    // declaration name

    let next = self.peek();
    if next.is_rcb() {
      return Ok(None);
    }
    self.asset_eof(next)?;

    if !next.is_ident() {
      return Err(Box::new(SyntaxError::from(
        SyntaxErrorKind::InvalidDeclName,
      )));
    }

    let mut decl_node = DeclarationNode::default();
    decl_node.name = next.to_string();
    decl_node.name_tokens.push(next);
    decl_node.loc.start = next.start_pos;

    self.consume();
    self.skip_ws_and_comments();

    // :

    let next = self.peek();
    if !next.is_colon() {
      return Err(Box::new(SyntaxError::from(SyntaxErrorKind::MissingColon)));
    }
    self.asset_eof(next)?;
    self.consume();
    self.skip_ws_and_comments();

    // parse value
    decl_node.value = self.parse_value()?;

    // check !important
    let mut rev_valus = decl_node.value.iter().rev();
    let last_value = rev_valus.next();
    let second_last_value = rev_valus.next();
    if second_last_value.is_some() && last_value.is_some() {
      let last_value = unsafe { last_value.unwrap_unchecked() };
      let second_last_value = unsafe { second_last_value.unwrap_unchecked() };
      if let Value::Operator(node) = second_last_value {
        if node.borrow().value.as_bytes() == b"!" {
          if let Value::Ident(node) = last_value {
            if node.borrow().name.as_bytes() == b"important" {
              decl_node.important = true;
            }
          }
        }
      }
    }

    Ok(Some(decl_node))
  }

  fn parse_value(&self) -> ParserResult<Vec<Value>> {
    let mut values: Vec<Value> = Vec::new();

    loop {
      let token = self.peek();
      self.asset_eof(token)?;

      if token.is_semi_colon() {
        self.consume();
        break;
      } else if token.is_rcb() {
        break;
      }

      self.consume();

      if !token.is_comment() {
        if token.is_ident() {
          let mut value_node = IdentNode::default();
          value_node.loc.start = token.start_pos;
          value_node.loc.end = token.end_pos;
          value_node.name = token.to_string();
          values.push(Value::Ident(Rc::new(RefCell::new(value_node))));
          self.skip_ws_and_comments();
        } else if token.is_string() {
          let mut value_node = StringNode::default();
          value_node.loc.start = token.start_pos;
          value_node.loc.end = token.end_pos;
          value_node.value = token.to_string();
          values.push(Value::String(Rc::new(RefCell::new(value_node))));
          self.skip_ws_and_comments();
        } else if token.is_url() {
          let mut value_node = URLNode::default();
          value_node.loc.start = token.start_pos;
          value_node.loc.end = token.end_pos;
          value_node.value = token.to_string();
          values.push(Value::URL(Rc::new(RefCell::new(value_node))));
          // consume RightParentheses
          self.consume();
          self.skip_ws_and_comments();
        } else if token.is_delim_loose() || token.is_comma() {
          let mut value_node = OperatorNode::default();
          value_node.loc.start = token.start_pos;
          value_node.loc.end = token.end_pos;
          value_node.value = token.to_string();
          values.push(Value::Operator(Rc::new(RefCell::new(value_node))));
          self.skip_ws_and_comments();
        } else if token.is_number() {
          let mut value_node = NumberNode::default();
          value_node.loc.start = token.start_pos;
          value_node.loc.end = token.end_pos;
          value_node.value = token.to_string();
          values.push(Value::Number(Rc::new(RefCell::new(value_node))));
          self.skip_ws_and_comments();
        } else if token.is_percentage() {
          let mut value_node = PercentageNode::default();
          value_node.loc.start = token.start_pos;
          value_node.loc.end = token.end_pos;
          value_node.value = token.to_string();
          values.push(Value::Percentage(Rc::new(RefCell::new(value_node))));
          self.skip_ws_and_comments();
        } else if token.is_dimension() {
          let mut value_node = DimensionNode::default();
          value_node.loc.start = token.start_pos;
          value_node.value = token.to_string();

          // unit
          let token = self.peek();
          debug_assert!(token.is_ident());
          value_node.unit = token.to_string();
          self.consume();

          value_node.loc.end = token.end_pos;

          values.push(Value::Dimension(Rc::new(RefCell::new(value_node))));
          self.skip_ws_and_comments();
        } else if token.is_function() {
          // Function
          values.push(Value::Function(Rc::new(RefCell::new(
            self.parse_function(token.to_string())?,
          ))));
          self.skip_ws_and_comments();
        } else {
          if token.is_lp() {
            // push stack
            self.value_fncall_stack.borrow_mut().push(0)
          } else if token.is_rp() {
            let mut stack = self.value_fncall_stack.borrow_mut();
            let stack_top = stack.pop();
            match stack_top {
              Some(1) => {
                // the end of function call
                // indicates that we are currently parsing the children of a function
                break;
              }
              _ => {}
            }
          }

          // TODO: Here we can do further analysis
          // Raw
          let mut value_node = RawNode::default();
          value_node.loc.start = token.start_pos;
          value_node.loc.end = token.end_pos;
          value_node.value = token.to_string();
          values.push(Value::Raw(Rc::new(RefCell::new(value_node))));
          self.skip_ws_and_comments();
        }
      }
    }

    Ok(values)
  }

  fn parse_function(&self, fn_name: String) -> ParserResult<FunctionNode> {
    let mut fn_node = FunctionNode::default();

    // function name
    fn_node.name = fn_name;

    let token = self.peek();
    debug_assert!(token.is_lp());
    self.value_fncall_stack.borrow_mut().push(1);
    self.consume();

    // function children
    fn_node.children = self.parse_value()?;

    Ok(fn_node)
  }

  fn asset_eof(&self, token: &Token) -> ParserResult<()> {
    if token.is_eof() {
      return Err(Box::new(SyntaxError::from(SyntaxErrorKind::UnexpectedEOF)));
    }
    Ok(())
  }

  fn skip_ws_and_comments(&self) {
    loop {
      let token = self.peek();
      if token.is_ws() || token.is_comment() {
        self.consume();
      } else {
        break;
      }
    }
  }
}
