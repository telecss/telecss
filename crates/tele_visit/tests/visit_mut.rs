use insta::assert_debug_snapshot;
use std::{cell::RefCell, rc::Rc};
use tele_parser::{DeclarationNode, FunctionNode, Parser};
use tele_tokenizer::Tokenizer;
use tele_visit::{MutVisitor, VisitMut};

struct Renamer;
impl<'s> MutVisitor<'s> for Renamer {
  fn visit_decl_node(&self, decl_node: Rc<RefCell<DeclarationNode<'s>>>) {
    let mut decl_node = decl_node.borrow_mut();
    decl_node.name = format!("-moz-{}", decl_node.name);
  }
}

struct ValueConverter;
impl<'s> MutVisitor<'s> for ValueConverter {
  fn visit_fn_node(&self, fn_node: Rc<RefCell<FunctionNode>>) {
    let mut fn_node = fn_node.borrow_mut();
    if fn_node.name == "v-bind" {
      fn_node.name = "var".to_owned();
    }
  }
}

#[test]
fn test_visit_decl_node() {
  // tokenizing
  let mut tokenizer: Tokenizer = r"  .foo { color: red; }  ".into();
  let tokens = tokenizer.tokenize().unwrap();
  // parsing
  let parser = Parser::from(tokens);
  let ast = parser.parse().unwrap();
  // transforming
  let transformer = VisitMut::new(Rc::clone(&ast), vec![Box::new(Renamer)]);
  transformer.transform();

  assert_debug_snapshot!(ast);
}

#[test]
fn test_multi_visitors() {
  // tokenizing
  let mut tokenizer: Tokenizer = r"  .foo { color: v-bind(clr); }  ".into();
  let tokens = tokenizer.tokenize().unwrap();
  // parsing
  let parser = Parser::from(tokens);
  let ast = parser.parse().unwrap();
  // transforming
  let transformer = VisitMut::new(
    Rc::clone(&ast),
    vec![Box::new(Renamer), Box::new(ValueConverter)],
  );
  transformer.transform();

  assert_debug_snapshot!(ast);
}
