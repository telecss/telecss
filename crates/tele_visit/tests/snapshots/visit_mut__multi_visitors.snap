---
source: crates/tele_visit/tests/visit_mut.rs
assertion_line: 55
expression: ast

---
RefCell {
    value: StyleSheetNode {
        loc: Loc {
            start: Pos {
                offset: 0,
                line: 1,
                column: 1,
            },
            end: Pos {
                offset: 32,
                line: 1,
                column: 33,
            },
        },
        statements: [
            RuleSet(
                RefCell {
                    value: RuleSetNode {
                        loc: Loc {
                            start: Pos {
                                offset: 2,
                                line: 1,
                                column: 3,
                            },
                            end: Pos {
                                offset: 30,
                                line: 1,
                                column: 31,
                            },
                        },
                        prelude: ".foo",
                        prelude_tokens: [
                            Token {
                                token_type: Delim,
                                start_pos: Pos {
                                    offset: 2,
                                    line: 1,
                                    column: 3,
                                },
                                end_pos: Pos {
                                    offset: 3,
                                    line: 1,
                                    column: 4,
                                },
                                content: ".",
                            },
                            Token {
                                token_type: Ident,
                                start_pos: Pos {
                                    offset: 3,
                                    line: 1,
                                    column: 4,
                                },
                                end_pos: Pos {
                                    offset: 6,
                                    line: 1,
                                    column: 7,
                                },
                                content: "foo",
                            },
                            Token {
                                token_type: WhiteSpace,
                                start_pos: Pos {
                                    offset: 6,
                                    line: 1,
                                    column: 7,
                                },
                                end_pos: Pos {
                                    offset: 7,
                                    line: 1,
                                    column: 8,
                                },
                                content: " ",
                            },
                        ],
                        declarations: [
                            RefCell {
                                value: DeclarationNode {
                                    loc: Loc {
                                        start: Pos {
                                            offset: 9,
                                            line: 1,
                                            column: 10,
                                        },
                                        end: Pos {
                                            offset: 0,
                                            line: 1,
                                            column: 1,
                                        },
                                    },
                                    name: "-moz-color",
                                    name_tokens: [
                                        Token {
                                            token_type: Ident,
                                            start_pos: Pos {
                                                offset: 9,
                                                line: 1,
                                                column: 10,
                                            },
                                            end_pos: Pos {
                                                offset: 14,
                                                line: 1,
                                                column: 15,
                                            },
                                            content: "color",
                                        },
                                    ],
                                    value: [
                                        Function(
                                            RefCell {
                                                value: FunctionNode {
                                                    loc: Loc {
                                                        start: Pos {
                                                            offset: 0,
                                                            line: 1,
                                                            column: 1,
                                                        },
                                                        end: Pos {
                                                            offset: 0,
                                                            line: 1,
                                                            column: 1,
                                                        },
                                                    },
                                                    name: "var",
                                                    children: [
                                                        Ident(
                                                            RefCell {
                                                                value: IdentNode {
                                                                    loc: Loc {
                                                                        start: Pos {
                                                                            offset: 23,
                                                                            line: 1,
                                                                            column: 24,
                                                                        },
                                                                        end: Pos {
                                                                            offset: 26,
                                                                            line: 1,
                                                                            column: 27,
                                                                        },
                                                                    },
                                                                    name: "clr",
                                                                },
                                                            },
                                                        ),
                                                    ],
                                                },
                                            },
                                        ),
                                    ],
                                    value_tokens: [],
                                    important: false,
                                },
                            },
                        ],
                    },
                },
            ),
        ],
    },
}
